### PR TITLE
Fix segfaults on close and gc

### LINKED ISSUE #157: Segfault on close with pending put
I am getting a segfault when I run the following unit test.  The segfault doesn't happen every time I run the test.  Any ideas?

``` javascript
var test        = require('tape');
var level       = require('level');
var path        = require('path');
var through2    = require('through2');
var JSONStream  = require('JSONStream');

function push(db) {

    return through2.obj(function(levelRequest, enc, cb) {

        var self = this;

        db.put(levelRequest.key, levelRequest.value, {sync:true}, function(error) {
            if (error) {
                console.log('encountered an error while putting ' + JSON.stringify(levelRequest) + ' on the database: '
                            + error);
                self.push({result:'error', key: levelRequest.key, msg: error});
            } else {
                self.push({result:'success', key: levelRequest.key});
            }
            cb();
        });

    })

}

test('can push values through a level database', function(t) {
    var db   = level('./test.db');

    var dbify = push(db);
    var stringify = JSONStream.stringify(false);

    dbify.pipe(stringify).pipe(process.stdout);

    dbify.write({key:'hi', value:"wisconsin"});

    db.close(function(er) {
        if (er) throw er;
        t.end();
    })

})
```

### LINKED ISSUE #589: Investigate segfault
See https://github.com/Level/leveldown/pull/587#issuecomment-457966267

### PR DESCRIPTION
**Pull request by vweevers**
Closes #589, closes #157.

### CODE REVIEW

**Comment by vweevers** [comment]
@ralphtheninja this is what a draft PR looks like :)

File: binding.cc [hunk 1]
```diff
@@ -813,11 +823,15 @@ NAPI_METHOD(db_close) {
   napi_value noop;
   napi_create_function(env, NULL, 0, noop_callback, NULL, &noop);
 
-  std::map< uint32_t, Iterator * >::iterator it;
-  for (it = database->iterators_.begin();
-       it != database->iterators_.end(); ++it) {
-    Iterator *iterator = it->second;
-    if (!iterator->ended_) {
```
Comment (vweevers) [inline binding.cc]: This `(!iterator->ended_)` check was redundant, we do the same in `iterator_end_do`

**Comment by vweevers** [comment]
[Node 10 on Linux](https://travis-ci.org/Level/leveldown/jobs/497733668) failed:

```
# test non-ended iterator
ok 851 no error from open()
ok 852 no error from batch()
ok 853 no error from next()
ok 854 correct key
ok 855 correct value
ok 856 no error from done()
node: ../deps/leveldb/leveldb-1.20/db/version_set.cc:798: leveldb::VersionSet::~VersionSet(): Assertion `dummy_versions_.next_ == &dummy_versions_' failed.
```

I'll see if I can replicate it.

**Comment by vweevers** [comment]
If I repeat the `test non-ended iterator` test lots of times I eventually hit [this node.js assertion](https://github.com/nodejs/node/blob/7e0ddf66b9e09a27e0761b487342027ed6284027/src/env-inl.h#L390) but not the LevelDB assertion.

**Comment by vweevers** [comment]
lion_king_rafiki_throw

**Comment by ralphtheninja** [comment]
Aaah now I see what draft means. Nice feature.

**Comment by vweevers** [comment]
This still occasionally fails. The current solution is an open invitation to race issues.

@ralphtheninja Instead, could we somehow prevent GC of `Iterator` objects? Until they are ended (by way of `it.end()` of `db.close()`). Ideally, the `EndWorker` is the only code path that removes the `Iterator` from the map. To achieve that, we must ensure that an `Iterator` is not destroyed before then.

File: binding.cc [hunk 2]
```diff
@@ -479,6 +489,7 @@ struct Iterator {
   }
 
   ~Iterator () {
+    assert(database_->closing_); // NOTE: debugging
```
Comment (vweevers) [inline binding.cc]: To clarify https://github.com/Level/leveldown/pull/597#issuecomment-466811204: this should always be true, currently isn't.

Comment (vweevers) [reply binding.cc]: No. Nevermind. I need a break..

**Comment by vweevers** [comment]
Note to self (I'll pick this up again end of week): [`napi_create_reference`](https://nodejs.org/api/n-api.html#n_api_napi_create_reference) should do the trick. Keep the reference count at 1, until the iterator is ended.

File: binding.cc [hunk 3]
```diff
@@ -506,6 +520,7 @@ struct Iterator {
     if (gte_ != NULL) {
       delete gte_;
     }
+    delete options_;
```
Comment (vweevers) [inline binding.cc]: Without this, `node --expose-gc test/leak-tester-iterators` shows endlessly growing memory, also on master.

File: binding.cc [hunk 4]
```diff
@@ -404,8 +406,17 @@ struct Database {
...
-  void ReleaseIterator (uint32_t id) {
-    iterators_.erase(id);
+  void AttachIterator (uint32_t id, Iterator* iterator) {
+    iterators_[id] = iterator;
+  }
+
+  void DetachIterator (uint32_t id) {
+    {
+      // Prevent parallel erase() calls
+      std::lock_guard<std::mutex> lock(iteratorsMutex_);
```
Comment (vweevers) [inline binding.cc]: Might be safe to remove now (it's possible that the problem was not with parallel iterators, but with parallel ending and gc on a single iterator).

File: binding.cc [hunk 5]
```diff
@@ -472,13 +484,15 @@ struct Iterator {
...
-      endWorker_(NULL) {
+      endWorker_(NULL),
+      ref_(NULL) {
     options_ = new leveldb::ReadOptions();
     options_->fill_cache = fillCache;
     options_->snapshot = database->NewSnapshot();
   }
 
   ~Iterator () {
+    assert(ended_); // NOTE: debugging
```
Comment (vweevers) [inline binding.cc]: @ralphtheninja can/should we keep this?

**Comment by vweevers** [comment]
The last commit (https://github.com/Level/leveldown/pull/597/commits/f8d97b9790eb241df6b50e5d13dd96e39c8de75b) is probably too much for this PR (because it is not related to iterators, but fixes #157) - it was something I wanted to explore. I'll revert it later (as well as squash the other commits).

**Comment by chjj** [comment]
@vweevers @ralphtheninja -- one thought I've had for maybe the past couple years now is that we could "auto end" iterators on GC.

This probably belongs in a separate issue, but it would allow for nice clean code with a promise wrapper. e.g.

``` js
for await (const [key, value] of db.iterator(options)) 
  console.log([key, value]);
```

As opposed to:

``` js
const iter = db.iterator(options);
try {
  for await (const [key, value] of iter)
    console.log([key, value]);
} finally {
  await iter.end();
}
```

The former is currently _impossible_ to do since the body of the loop may throw an error, which the async iterator cannot account for.

What do you guys think?

**Comment by ralphtheninja** [comment]
It's a good idea.

**Comment by vweevers** [comment]
I don't think the lack of syntactic sugar is a decisive argument. Note there is an [ECMAScript proposal for explicit resource management](https://github.com/tc39/proposal-using-statement). That said, I don't have a good argument for preserving `.end()` other than a slight personal preference for knowing and controlling the resources your app uses.

If we can make it happen, great!

**Comment by vweevers** [comment]
While this part of `leveldown` is still fresh in my head, I'll play around a bit.

**Comment by vweevers** [comment]
I think I have a viable, lock-free strategy:

- Make `iterator.end()` a noop in JS-land
- Which means `EndWorker` can only be called on `db.close()`, so remove `EndWorker`, instead do that work in `CloseWorker`
- Make NextWorker a PriorityWorker, so that it defers closing (instead of deferring EndWorker)
- Maintain a reference count per iterator, initially 0
- `iterator.next()` increases the refcount to prevent GC while nexting
- `db.close()` increases the refcounts of all iterators, to prevent GC until CloseWorker executes.

**Pull request merged**